shakenfist/blob.py

```diff
@@ -193,6 +203,58 @@ def ref_count_dec(self):
 
             return new_count
 
+    def ensure_local(self, locks):
+        blob_dir = os.path.join(config.STORAGE_PATH, 'blobs')
+        blob_path = os.path.join(blob_dir, self.uuid)
+        os.makedirs(blob_dir, exist_ok=True)
+
+        if os.path.exists(blob_path):
+            return
+
+        locations = self.locations
+        random.shuffle(locations)
+        blob_source = locations[0]
+
+        url = 'http://%s:%d/blob/%s' % (blob_source, config.API_PORT,
+                                        self.uuid)
+        admin_token = util_general.get_api_token(
+            'http://%s:%d' % (blob_source, config.API_PORT))
+        r = requests.request('GET', url, stream=True,
+                             headers={'Authorization': admin_token,
+                                      'User-Agent': util_general.get_user_agent()})
+
+        with open(blob_path + '.partial', 'wb') as f:
+            last_refresh = 0
+
+            for chunk in r.iter_content(chunk_size=8192):
+                f.write(chunk)
+
+                if time.time() - last_refresh > LOCK_REFRESH_SECONDS:
+                    db.refresh_locks(locks)
+                    last_refresh = time.time()
+
+        os.rename(blob_path + '.partial', blob_path)
+        self.observe()
+
+    def request_replication(self):
+        targets = config.BLOB_REPLICATION_FACTOR - len(self.locations)
+        if targets > 0:
+            by_disk = {}
+            for n in Nodes([node_active_states_filter]):
+                metrics = db.get_metrics(n.fqdn)
+                disk_free_gb = int(metrics.get('disk_free', '0')) / GIGABYTE
+                by_disk.setdefault(disk_free_gb, [])
+                by_disk[disk_free_gb].append(n.fqdn)
```

Comment:
Should this check for minimum space available? Or better, check for minimum space after the blob has been replicated?

Reply:
I've had a go at a tweak to do this.

Reply:
I've tweaked the matching place in the scheduler too.